doc/modules/compiler/utils.rst

```diff
@@ -181,16 +181,86 @@ functions with :ref:`mux-kernel <specifications/mux-compiler-spec:Function
+ascending order from `0` to `N-1` through the innermost `X` dimension, followed
+by the `Y` dimension, and lastly the `Z` dimension.
+
+Conceptually, the pass transforms ``old_kernel`` into ``new_kernel`` in the
+example below:
+
+.. code:: cpp
+
+   void old_kernel(int *in, int *out) {
+     size_t id = get_global_id(0);
+     out[id] = in[id] * 4;
+   }
+
+   void new_kernel(int *in, int *out) {
+     for (size_t z = 0; z != get_local_size(2); z++) {
+       for (size_t y = 0, y != get_local_size(1); y++) {
+         for (size_t x = 0, x != get_local_size(0); x++) {
+           size_t id = get_global_id(0);
+           out[id] = in[id] * 4;
+         }
+       }
+     }
+   }
+
+
+To satisfy the programming model, the pass must be careful around control
+barriers and *barrier-like* functions. The ``WorkItemLoopsPass`` splits a
+kernel into separately executing kernel functions using barrier calls as
+boundaries.
+
+.. code:: cpp
+
+   void old_kernel(int *in, int *out) {
+     size_t id = get_global_id(0);
+     out[id * 2] = in[id];
+     // All work-items in the work-group must encounter the barrier before any
+     // are allowed to continue execution beyond the barrier.
+     work_group_barrier(CLK_GLOBAL_MEM_FENCE);
+     out[id * 2 + 1] = in[id] * 4;
+   }
+
+   void new_kernel(int *in, int *out) {
+     for (size_t z = 0; z != get_local_size(2); z++) {
+       for (size_t y = 0, y != get_local_size(1); y++) {
+         for (size_t x = 0, x != get_local_size(0); x++) {
+           size_t id = get_global_id(0);
+           out[id * 2] = in[id];
+         }
+       }
+     }
+
+     // The control aspect of the barrier has been satisfied by the loops, so
+     // it has been decomposed to just a memory barrier.
+     mem_fence(CLK_GLOBAL_MEM_FENCE);
+
+     for (size_t z = 0; z != get_local_size(2); z++) {
+       for (size_t y = 0, y != get_local_size(1); y++) {
+         for (size_t x = 0, x != get_local_size(0); x++) {
+           size_t id = get_global_id(0);
+           out[id * 2 + 1] = in[id] * 4;
+         }
+       }
+     }
+   }
 
-The ``HandleBarriersPass`` splits a kernel into separately executing kernel
-functions using barrier calls as boundaries. To propagate data dependencies
-between these functions an analysis is done to create a struct of live variables
-which is passed as an argument to each kernel. Generated kernels then reference
-this struct rather than the original values. Work-item scheduling is also redone
-so that every work-item in a work-group invokes each kernel function before
-advancing.
+To propagate data dependencies between these functions an analysis
+is done to create a struct of live variables which is passed as an argument to
+each kernel. Generated kernels then reference this struct rather than the
+original values. Work-item scheduling is also redone so that every work-item in
+a work-group invokes each kernel function before advancing.
```

Comment:
I wonder if it's worth mention that this is a vector across the workgroup for non-uniform values, since this is a source of confusion (or is it a vector of structs I can't remember)

I also seem to recall a different live struct for vector and scalar parts